datasafe-storage/datasafe-storage-api/src/main/java/de/adorsys/datasafe/storage/api/UserBasedDelegatingStorage.java

```diff
@@ -0,0 +1,33 @@
+package de.adorsys.datasafe.storage.api;
+
+import de.adorsys.datasafe.types.api.resource.AbsoluteLocation;
+import de.adorsys.datasafe.types.api.resource.Uri;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+public class UserBasedDelegatingStorage extends BaseDelegatingStorage {
+    private final Map<String, StorageService> clientByBucket = new ConcurrentHashMap<>();
+    private final String amazonBucket;
+
+    public UserBasedDelegatingStorage(Function<String, StorageService> storageServiceBuilder, String amazonBucket) {
+        this.storageServiceBuilder = storageServiceBuilder;
+        this.amazonBucket = amazonBucket;
+    }
+
+    // Builder to create S3 or other kind of Storage service
+    private final Function<String, StorageService> storageServiceBuilder;
+
+    @Override
+    protected StorageService service(AbsoluteLocation location) {
+        Uri uri = location.location();
+        String[] parts = uri.getPath().replaceAll("^/", "").split("/");
```

Comment:
Maybe it can be done simpler:
```
Pattern userPattern = Pattern.compile(".+/user-(\\d+)/.+");
Matcher matcher = userPattern.matcher(location.location().asString())
matcher.matches();
int userNum = Integer.valueOf(matcher.group(1));
return clientByBucket.computeIfAbsent(buckets[userNum % buckets.length], storageServiceBuilder);
```
```

Reply:
ok